hoomd/hpmc/tune/nec_chain_time.py

```diff
@@ -0,0 +1,286 @@
+# This file is part of the HOOMD-blue project, released under the BSD 3-Clause
+# License.
+
+"""Implement ChainTime."""
+
+from hoomd.custom import _InternalAction
+from hoomd.data.parameterdicts import ParameterDict
+from hoomd.data.typeconverter import OnlyTypes
+from hoomd.tune import _InternalCustomTuner
+from hoomd.tune.attr_tuner import (_TuneDefinition, SolverStep, ScaleSolver,
+                                   SecantSolver)
+from hoomd.hpmc.integrate_nec import HPMCNECIntegrator
+
+
+class _ChainTimeTuneDefinition(_TuneDefinition):
+    """Encapsulates getting particles per chain and getting/setting chain time.
+
+    This class should only be used for the _InternalChainTime class to tune
+    HPMC-NEC chain time. For this class 'x' is the chain time and 'y' is the
+    number of particles per chain.
+    """
+
+    def __init__(self, target, domain=None):
+        self.integrator = None
+        self.previous_hit = None
+        self.previous_start = None
+        self.previous_particles_per_chain = None
+        super().__init__(target, domain)
+
+    def _get_y(self):
+        statistics = getattr(self.integrator, "nec_counters")
+        chain_hit = statistics.chain_at_collision_count
+        chain_start = statistics.chain_start_count
+
+        # We return None when no chains are recorded since we don't want
+        # the chain time to be updated. Likewise, when we do not have a
+        # previous recorded number of particles in a chain we return None
+        # since what happened previous timesteps may not be indicative of
+        # the current system.
+        #
+        # None in the hoomd solver infrastructure means that the value
+        # either cannot be computed or would be inaccurate at the current
+        # time. It informs the `SolverStep` object to skip tuning this
+        # attribute for now.
+        if self.previous_start is None or chain_start == 0:
+            self.previous_hit = chain_hit
+            self.previous_start = chain_start
+            return None
+
+        # If no more chains have been recorded return previous
+        # number of particles_per_chain per chain.
+        elif self.previous_start == chain_start:
+            return self.previous_particles_per_chain
+
+        # If we have recorded a previous total larger than the current one
+        # then this condition implies a new run call. We should be able to
+        # tune here as we have no other indication the system has changed.
+        elif ((self.previous_start > chain_start)
+              or (self.previous_hit > chain_hit)):
+            particles_per_chain = chain_hit / chain_start
+        else:
+            # flake8 does not allow a single-line calculation
+            # W504 line break after binary operator
+            delta_prev = chain_hit - self.previous_hit
+            delta_start = chain_start - self.previous_start
+            particles_per_chain = delta_prev / delta_start
```

Comment:
`delta_prev` doesn't make much sense `delta_hit` would be better. Also, you can use parenthesis to enable multi-line mathematical expressions.